tower-http/src/follow_redirect/mod.rs

```diff
@@ -0,0 +1,325 @@
+//! Middleware for following redirections.
+
+pub mod policy;
+
+use self::policy::{ActionKind, Attempt, Policy, Standard};
+use futures_core::ready;
+use http::{
+    header::LOCATION, HeaderMap, HeaderValue, Method, Request, Response, StatusCode, Uri, Version,
+};
+use http_body::Body;
+use iri_string::{
+    spec::UriSpec,
+    types::{RiAbsoluteString, RiReferenceStr},
+};
+use pin_project::pin_project;
+use std::{
+    convert::TryFrom,
+    future::Future,
+    mem,
+    pin::Pin,
+    str,
+    task::{Context, Poll},
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// [`Layer`] for retrying requests with a [`Service`] to follow redirection responses.
+#[derive(Clone, Copy, Debug, Default)]
+pub struct FollowRedirectLayer<P = Standard> {
+    policy: P,
+}
+
+impl FollowRedirectLayer {
+    /// Create a new [`FollowRedirectLayer`] with a [`Standard`] redirection policy.
+    pub fn standard() -> Self {
+        Self::default()
+    }
+}
+
+impl<P> FollowRedirectLayer<P> {
+    /// Create a new [`FollowRedirectLayer`] with the given redirection [`Policy`].
+    pub fn new(policy: P) -> Self {
+        FollowRedirectLayer { policy }
+    }
+}
+
+impl<S, P> Layer<S> for FollowRedirectLayer<P>
+where
+    S: Clone,
+    P: Clone,
+{
+    type Service = FollowRedirect<S, P>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        FollowRedirect::new(inner, self.policy.clone())
+    }
+}
+
+/// Middleware that retries requests with a [`Service`] to follow redirection responses.
+#[derive(Clone, Copy, Debug)]
+pub struct FollowRedirect<S, P = Standard> {
+    inner: S,
+    policy: P,
+}
+
+impl<S> FollowRedirect<S> {
+    /// Create a new [`FollowRedirect`] with a [`Standard`] redirection policy.
+    pub fn standard(inner: S) -> Self {
+        Self::new(inner, Standard::default())
+    }
+}
+
+impl<S, P> FollowRedirect<S, P>
+where
+    P: Clone,
+{
+    /// Create a new [`FollowRedirect`] with the given redirection [`Policy`].
+    pub fn new(inner: S, policy: P) -> Self {
+        FollowRedirect { inner, policy }
+    }
+
+    /// Returns a new [`Layer`] that wraps services with a `FollowRedirect` middleware.
+    ///
+    /// [`Layer`]: tower_layer::Layer
+    pub fn layer(policy: P) -> FollowRedirectLayer<P> {
+        FollowRedirectLayer::new(policy)
+    }
+}
+
+impl<ReqBody, ResBody, S, P> Service<Request<ReqBody>> for FollowRedirect<S, P>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>> + Clone,
+    ReqBody: Body + Default,
+    P: Policy<ReqBody, S::Error> + Clone,
+{
+    type Response = Response<ResBody>;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future, S, ReqBody, P>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: Request<ReqBody>) -> Self::Future {
+        let mut service = self.inner.clone();
```

Comment:
The clone of a ready service might not be ready. This is a workaround:

```suggestion
        let clone = self.inner.clone();
        let mut service = std::mem::replace(&mut self.inner, clone);
```

More details [here](https://github.com/tower-rs/tower/issues/547#issuecomment-767629149).